pkg/pipeline/providers/gitlab/actions.go

```diff
@@ -0,0 +1,342 @@
+		return nil, errors.New("unauth")
+	}
+
+	remote, err := remote.New(config)
+	if err != nil {
+		return nil, err
+	}
+	account, err := remote.Login(code)
+	if err != nil {
+		return nil, err
+	}
+	account.Name = strings.ToLower(fmt.Sprintf("%s-%s-%s", config.Namespace, model.GitlabType, account.Spec.LoginName))
+	account.Namespace = userID
+	account.Spec.UserName = userID
+	account.Spec.ProjectName = config.ProjectName
+	_, err = g.SourceCodeCredentials.Create(account)
+	if apierror.IsAlreadyExists(err) {
+		exist, err := g.SourceCodeCredentialLister.Get(userID, account.Name)
+		if err != nil {
+			return nil, err
+		}
+		account.ResourceVersion = exist.ResourceVersion
+		return g.SourceCodeCredentials.Update(account)
+	} else if err != nil {
+		return nil, err
+	}
+	return account, nil
+}
+
+func (g *GlProvider) disableAction(request *types.APIContext) error {
+
+	ns, _ := ref.Parse(request.ID)
+	o, err := g.SourceCodeProviderConfigs.ObjectClient().UnstructuredClient().GetNamespaced(ns, model.GitlabType, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	u, _ := o.(runtime.Unstructured)
+	config := u.UnstructuredContent()
+	if e, ok := config[client.SourceCodeProviderConfigFieldEnabled].(bool); ok && e {
+		config[client.SourceCodeProviderConfigFieldEnabled] = false
+		if _, err := g.SourceCodeProviderConfigs.ObjectClient().Update(model.GitlabType, o); err != nil {
+			return err
+		}
+		if t, ok := config["projectName"].(string); ok && t != "" {
+			return g.cleanup(config["projectName"].(string))
+		}
+	}
+
+	return nil
+}
+
+func (g *GlProvider) cleanup(projectID string) error {
+
+	pipelines, err := g.PipelineLister.List("", labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, p := range pipelines {
+		if p.Spec.ProjectName != projectID {
+			continue
+		}
+		if err := g.Pipelines.DeleteNamespaced(p.Namespace, p.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+
+	pipelineExecutions, err := g.PipelineExecutionLister.List("", labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, e := range pipelineExecutions {
+		if e.Spec.ProjectName != projectID {
+			continue
+		}
+		if err := g.PipelineExecutions.DeleteNamespaced(e.Namespace, e.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+
+	credentials, err := g.SourceCodeCredentialLister.List("", labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, credential := range credentials {
+		if credential.Spec.SourceCodeType != model.GitlabType || credential.Spec.ProjectName != projectID {
+			continue
+		}
+		if err := g.SourceCodeCredentials.DeleteNamespaced(credential.Namespace, credential.Name, &metav1.DeleteOptions{}); err != nil {
+			return err
+		}
+	}
+
+	repositories, err := g.SourceCodeRepositoryLister.List("", labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, repo := range repositories {
+		if repo.Spec.SourceCodeType != model.GitlabType || repo.Spec.ProjectName != projectID {
```

Comment:
indexer :) In addition to project, you can add an indexer for the sourceCodeType